Lib/_pylong.py

```diff
@@ -82,7 +86,56 @@ def inner(n, w):
 
 def int_to_decimal_string(n):
     """Asymptotically fast conversion of an 'int' to a decimal string."""
-    return str(int_to_decimal(n))
+    w = n.bit_length()
+    if w > 450_000 and _decimal is not None:
+        # It is only usable with the C decimal implementation.
+        # _pydecimal.py calls str() on very large integers, which in its
+        # turn calls int_to_decimal_string(), causing very deep recursion.
+        return str(int_to_decimal(n))
+
+    # Fallback algorithm for the case when the C decimal module isn't
+    # available.  This algorithm is asymptotically worse than the algorithm
+    # using the decimal module, but better than the quadratic time
+    # implementation in longobject.c.
+    def inner(n, w):
+        if w <= 1000:
+            try:
+                # In normal cases n is small enough here, so the built-in
+                # longobject.c algorithm is the fastest.
+                return str(n)
+            except ValueError:
+                # In very rare cases, when our guess of w is too small and
+                # n is too large, we can hit the limit for int to str
+                # conversion in str().  Call int_to_decimal_string() which
+                # has no such limitation directly.
+                return int_to_decimal_string(n)
```

Comment:
I am not actually sure that this is needed. If `int_to_decimal_string()` is only called from `str()`, it never hits the limit for int to str conversion. But if it is used independently, it could potentially raise ValueError in `str()`, although practically it is impossible on modern computers. And if you have exabytes of RAM and use Python to convert exabytes-long int to str, you can hire somebody to fix Python for you.

Reply:
I wouldn't bother with this. `_pylong` is deliberately not documented. As the comments say at the start:

> Functions provided by this module should be considered private and not part of any public API.

So "used independently" is not a supported use case.